src/fingerprint_bond.cpp

```diff
@@ -0,0 +1,917 @@
+					dfeaturesy[jj*f+count]+=dot1*c5;
+					dfeaturesz[jj*f+count]+=dot1*c6;
+					c4*=dot;
+					c5*=dot;
+					c6*=dot;
+					count++;
+					for (m=1;m<mb;m++){
+						double c7 = dot1*(m*c1+c4);
+						double c8 = dot1*(m*c2+c5);
+						double c9 = dot1*(m*c3+c6);
+						dfeaturesx[jj*f+count]+=c7;
+						dfeaturesy[jj*f+count]+=c8;
+						dfeaturesz[jj*f+count]+=c9;
+						dot1*=dot;
+						features[count++]+=dot1;
+					}
+				}
+			}
+		}
+		else {
+			for (kk=0;kk<jnum; kk++){
+				if (expr[kk][0]==0)continue;
+//				int k1 = jlist[kk];
+//				k1 &= NEIGHMASK;
+//				ktype = pair->map[type[k1]];
+				ktype = tn[kk];
+				if (ktypes != nelements && ktypes != ktype){
+					continue;
+				}
+				count = startingneuron;
+				double dot = (y[jj][0]*y[kk][0]+y[jj][1]*y[kk][1]+y[jj][2]*y[kk][2]);
+				double c1  = ri[jj]*(y[kk][0]-dot*y[jj][0]);
+				double c2  = ri[jj]*(y[kk][1]-dot*y[jj][1]);
+				double c3  = ri[jj]*(y[kk][2]-dot*y[jj][2]);
+				double c10 = ri[kk]*(y[jj][0]-dot*y[kk][0]);
+				double c11 = ri[kk]*(y[jj][1]-dot*y[kk][1]);
+				double c12 = ri[kk]*(y[jj][2]-dot*y[kk][2]);
+//				double c1 = 2*ri[jj]*y[kk][0]*(1-y[jj][0]*y[jj][0]);
+//				double c2 = 2*ri[jj]*y[kk][1]*(1-y[jj][1]*y[jj][1]);
+//				double c3 = 2*ri[jj]*y[kk][2]*(1-y[jj][2]*y[jj][2]);
+//				double c10 = 2*ri[kk]*y[jj][0]*(1-y[kk][0]*y[kk][0]);
+//				double c11 = 2*ri[kk]*y[jj][1]*(1-y[kk][1]*y[kk][1]);
+//				double c12 = 2*ri[kk]*y[jj][2]*(1-y[kk][2]*y[kk][2]);
+				for (n=0;n<kb;n++){
+					double dot1=expr[jj][n]*expr[kk][n];
+					double c4 = c41[n]/2;
+					double c5 = c51[n]/2;
+					double c6 = c61[n]/2;
+					double ct2 = -ct[n]/2+dfc[kk];
+					double c42 = ct2*y[kk][0];
+					double c52 = ct2*y[kk][1];
+					double c62 = ct2*y[kk][2];
+					//m=0
+					features[count]+=dot1;
+					dfeaturesx[jj*f+count]+=dot1*c4;
+					dfeaturesy[jj*f+count]+=dot1*c5;
+					dfeaturesz[jj*f+count]+=dot1*c6;
+					dfeaturesx[kk*f+count]+=dot1*c42;
+					dfeaturesy[kk*f+count]+=dot1*c52;
+					dfeaturesz[kk*f+count]+=dot1*c62;
+					c4*=dot;
+					c5*=dot;
+					c6*=dot;
+					c42*=dot;
+					c52*=dot;
+					c62*=dot;
+					count++;
+					for (m=1;m<mb;m++){
+						double c7 = dot1*(m*c1+c4);
+						double c8 = dot1*(m*c2+c5);
+						double c9 = dot1*(m*c3+c6);
+						dfeaturesx[jj*f+count]+=c7;
+						dfeaturesy[jj*f+count]+=c8;
+						dfeaturesz[jj*f+count]+=c9;
+						dfeaturesx[kk*f+count]+=dot1*(m*c10+c42);
+						dfeaturesy[kk*f+count]+=dot1*(m*c11+c52);
+						dfeaturesz[kk*f+count]+=dot1*(m*c12+c62);
+						dot1*=dot;
+						features[count++]+=dot1;
+					}
+				}
+			}
+		}
+	}
+	for (jj=0;jj<jnum;jj++){
+		if (expr[jj][0]==0){continue;}
+		count = startingneuron;
+		for (n=0;n<kb;n++){
+			for (m=0;m<mb;m++){
+				dfeaturesx[jnum*f+count]-=dfeaturesx[jj*f+count];
+				dfeaturesy[jnum*f+count]-=dfeaturesy[jj*f+count];
+				dfeaturesz[jnum*f+count]-=dfeaturesz[jj*f+count];
+				count++;
+			}
+		}
+	}
+}
+
+int Fingerprint_bond::factorial(int n) {
```

Comment:
we have an optimized `factorial()` function in the MathSpecial namespace. Please use that one.

Reply:
Done. I created a wrapper for it in pair_rann.